### PR TITLE
doc: extensions: add custom role 'module_file'

### PR DESCRIPTION
**Pull request by gautesl**
Added another custom role `module_file` which functions like `zephyr_file` but for any module in the west manifest.

The new role must take a module as an argument. Example syntax:

```rst
:module_file:`module-name: path/within/module`
```

or with custom link text:

```rst
:module_file:`my link text <module-name: path/within/module>`
```

Signed-off-by: Gaute Svanes Lunde <gaute.lunde@nordicsemi.no>

### CODE REVIEW

File: doc/_extensions/zephyr/link-roles.py [hunk 1]
```diff
@@ -21,61 +23,74 @@
...
 def get_github_rev():
     try:
-        output = subprocess.check_output('git describe --exact-match',
-                                         shell=True, stderr=subprocess.DEVNULL)
+        output = subprocess.check_output(
+            "git describe --exact-match", shell=True, stderr=subprocess.DEVNULL
+        )
     except subprocess.CalledProcessError:
-        return 'main'
+        return "main"
```
Comment (gmarull) [inline doc/_extensions/zephyr/link-roles.py:30]: please split formatting changes from actual extension additions

File: doc/_extensions/zephyr/link-roles.py [hunk 2]
```diff
@@ -21,61 +23,74 @@
...
-        except ValueError:
-            pass
-
-    # If the search failed, fall back on the mainline URL.
-    if baseurl is None:
-        baseurl = 'https://github.com/zephyrproject-rtos/zephyr'
-
-    app.add_role('zephyr_file', autolink('{}/blob/{}/%s'.format(baseurl, rev)))
-    app.add_role('zephyr_raw', autolink('{}/raw/{}/%s'.format(baseurl, rev)))
+    baseurl = "https://github.com/zephyrproject-rtos/zephyr"
```
Comment (gmarull) [inline doc/_extensions/zephyr/link-roles.py:36]: can't baseurl for Zephyr be determined from the manifest?

Comment (gautesl) [reply doc/_extensions/zephyr/link-roles.py:36]: Not if the manifest does not exist, which is the case for the Zephyr docs. This is only used by the already existing `zephyr_file` and `zephyr_raw` roles

Comment (gmarull) [reply doc/_extensions/zephyr/link-roles.py:36]: Hmm Zephyr has a manifest: https://github.com/zephyrproject-rtos/zephyr/blob/main/west.yml

Comment (gautesl) [reply doc/_extensions/zephyr/link-roles.py:36]: I must have built it wrong locally, I will look into it

File: doc/_extensions/zephyr/link-roles.py [hunk 3]
```diff
@@ -68,16 +47,49 @@ def setup(app):
...
+        # software with forks of the zephyr repository, and getting
+        # :zephyr_file: / :zephyr_raw: output that links to the fork,
+        # instead of mainline zephyr.
+        projects = [p.name for p in west_manifest.projects] if west_manifest else []
+        if module in projects:
+            project = west_manifest.get_projects([module])[0]
+            baseurl = project.url
+            rev = project.revision
+
+        if baseurl is None or module is None:
```
Comment (gmarull) [inline doc/_extensions/zephyr/link-roles.py:82]: can this condition ever hold true?

File: doc/conf.py [hunk 4]
```diff
@@ -251,6 +251,10 @@
 
 warnings_filter_config = str(ZEPHYR_BASE / "doc" / "known-warnings.txt")
 
+# -- Options for zephyr.link-roles ----------------------------------------
+
+link_roles_default_url = "https://github.com/zephyrproject-rtos/zephyr"
```
Comment (gmarull) [inline doc/conf.py:256]: I'd name it `link_roles_manifest_baseurl`, this should only be used by the repository containing the manifest.

File: doc/_extensions/zephyr/link-roles.py (hunk 3 above)
Comment (gautesl) [reply doc/_extensions/zephyr/link-roles.py:82]: `module` can be `None` if using the new role `:module_file:` without a module argument. 
`baseurl` can be `None` given the following circumstances:
- `:module_file:` is used while the `link_roles_default_url` config value is not set (defaults to `None`). 
- The provided module is not found in the manifest (invalid module)

Comment (gautesl) [reply doc/_extensions/zephyr/link-roles.py]: Failure tests rewritten to include the case where both a default url and an invalid module are provided

File: doc/conf.py (hunk 4 above)
Comment (gautesl) [reply doc/conf.py]: Fixed. Also introduced `link_roles_manifest_project` to get the name of the manifest project.

**Comment by gautesl** [comment]
Added another configuration value `link_roles_manifest_project` to properly check for invalid modules. Renamed `link_roles_default_url` to `link_roles_manifest_baseurl`, removed potential spaces from provided modules and links and cleared up the error checking.

**Review by gmarull — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**